pkg/minikube/download/image.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2020 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package download
+
+import (
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/cheggaaa/pb/v3"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	"k8s.io/minikube/pkg/minikube/constants"
+	"k8s.io/minikube/pkg/minikube/localpath"
+)
+
+var (
+	defaultPlatform = v1.Platform{
+		Architecture: runtime.GOARCH,
+		OS:           "linux",
+	}
+)
+
+// imageExistsInCache if img exist in local cache directory
+func imageExistsInCache(img string) bool {
+	f := filepath.Join(constants.KICCacheDir, path.Base(img)+".tar")
+	f = localpath.SanitizeCacheDir(f)
+
+	// Check if image exists locally
+	klog.Infof("Checking for %s in local cache directory", img)
+	if st, err := os.Stat(f); err == nil {
+		if st.Size() > 0 {
+			klog.Infof("Found %s in local cache directory, skipping pull", img)
+			return true
+		}
+	}
+	// Else, pull it
+	return false
+}
+
+var checkImageExistsInCache = imageExistsInCache
+
+// imageExistsInDaemon if img exist in local docker daemon
+func imageExistsInDaemon(img string) bool {
+	// Check if image exists locally
+	klog.Infof("Checking for %s in local docker daemon", img)
+	cmd := exec.Command("docker", "images", "--format", "{{.Repository}}:{{.Tag}}@{{.Digest}}")
+	if output, err := cmd.Output(); err == nil {
+		if strings.Contains(string(output), img) {
+			klog.Infof("Found %s in local docker daemon, skipping pull", img)
+			return true
+		}
+	}
+	// Else, pull it
+	return false
+}
+
+var checkImageExistsInDaemon = imageExistsInDaemon
+
+// ImageToCache writes img to the local cache directory
+func ImageToCache(img string) error {
```

Comment:
Why `ImageToCache` is a better name than `WriteImageToCache`? :)

Reply:
It looks weird if `download` part is changed during import -  for example, with dot imports (` import . ".../download" ` bad idea) or if the package is given a new name - `import q "../download"`. Also we have `download.Tag(string)` for example, which doesn't actually download any tags :) So I'd rename `download.Binary` instead :)
Not a big deal though.

Reply:
I would say this is relatively minor issue. There are other places that already use the package name as a verb (ex. `delete.PossibleLeftOvers`), also renaming `download.Binary` to something meaningful is hard, since the best name for it `download.DownloadBinary` would trigger a stuttering lint. I would say the solution here if necessary would be to move/rename all non-conforming functions, but that's a separate PR.